pallets/gear/src/lib.rs

```diff
@@ -645,21 +692,37 @@ pub mod pallet {
                             max_gas_spent.max(initial_gas.saturating_sub(remaining_gas));
                     }
 
-                    if let JournalNote::MessageDispatched {
-                        outcome: CoreDispatchOutcome::MessageTrap { trap, .. },
-                        ..
-                    } = note
-                    {
-                        return Err(format!(
-                            "Program terminated with a trap: {}",
-                            trap.unwrap_or_else(|| "No reason".to_string())
-                        )
-                        .into_bytes());
+                    match note {
+                        JournalNote::SendDispatch { dispatch, .. } => {
+                            gas_to_send =
+                                gas_to_send.saturating_add(dispatch.gas_limit().unwrap_or(0));
+                        }
+
+                        JournalNote::GasBurned { amount, .. } => {
+                            burnt = burnt.saturating_add(amount);
+                        }
+
+                        JournalNote::MessageDispatched {
+                            outcome: CoreDispatchOutcome::MessageTrap { trap, .. },
+                            ..
+                        } => {
+                            return Err(format!(
+                                "Program terminated with a trap: {}",
```

Comment:
Safety flag for me should indicate do we allow other programs panic

For any case panic in currently checking contract means error and should be so
But for many other cases if our contract handles other contracts panic, this is the correct behaviour.

With safety flag we will or won't fail this fn if panics occurred not in main calculating program